##  **_Βαμβακάς Παναγιώτης_**
## ge17090

--------

# Ασκήσεις 35-37

In [1]:
import statistics as stat
from simple_markov_chain_lib import markov_chain

p = 1 / 6

# A dictionary for the initial distribution.
# We prescribe the initial distribution

init_dist = {1: 1.0}

# A dictionary for the transition probability  matrix.
# Every state-key corresponds to a list with tuples of (Next_State,Probability)
markov_table = {
    1: {2: 1.},
    2: {2: 2 / 3, 3: 1 / 3},
    3: {1: p, 2: 1 - p}
}

# Markov Chain construction
mc = markov_chain(markov_table, init_dist)

# Experiment parameters
N = 20000  # number of samples
steps = 20  # the target time
p_list = []  # the list of the phat estimates

for a in range(50):  # iteration to produce the estimates and then add them to our list
    counter = 0  # to count the number of times the event {X_20  = 1} occurs

    # Simulation
    for i in range(N):
        mc.start()  # new experiment
        for j in range(steps):
            mc.move()
        if mc.running_state == 1:
            counter += 1

    phat = counter / N
    p_list.append(phat)  # A new estimate of phat is added on the list

print(
    """ 
    The sample mean of phat is {0:.5f} and the sample variance is {1:.5f}
    """.format(stat.mean(p_list), stat.variance(p_list)))
# The code to compute the sample mean and variance of our phat list

 
    The sample mean of phat is 0.04076 and the sample variance is 0.00000
    


## Απαντήσεις ασκήσεων 35-37
    
  Για την υλοποίηση των ασκήσεων, δημιουργήσαμε μια λίστα , την `p_list`, τρέξαμε τον κώδικα για την εύρεση της εκτιμήτριας 50 φορές, και κάθε φορά, αποθηκεύαμε το αποτέλεσμα στη λίστα.

1. Με χρήση της εντολής `stat.mean(p_list)`, υπολογίσαμε τη δειγματική μέση τιμή για τις $M=50$ εκτιμήσεις Monte Carlo της πιθανότητας $\mathbb{P}_1[X_{20}=1]$ όταν $N=200$ και όταν $N=20.000$, και βρήκαμε: $\overline{\hat{p}_{200}}=0.03890$ και $\overline{\hat{p}_{20000}}=0.04001$
---
2. Άμα ακολουθήσουμε τον θεωρητικό τρόπο, δηλαδή να διαγωνοποιήσουμε τον πίνακα $P = 
\begin{pmatrix}
0 & 1 & 0\\
0 & 2/3 & 1/3\\
1/6 & 5/6 & 0
\end{pmatrix}$ 

, βρίσκουμε $P^{20} = \begin{pmatrix}
1 & \frac{12-18i}{13} & \frac{12+18i}{13} \\
1 & \frac{-5+i}{13} & \frac{-5-i}{13} \\
1 & 1 & 1
\end{pmatrix}
\begin{pmatrix}
1 & 0 & 0\\
0 & ({\frac{-1-i}{6}})^{20} & 0\\
0 & 0 & ({\frac{-1+i}{6}})^{20}
\end{pmatrix}
{\begin{pmatrix}
1 & \frac{12-18i}{13} & \frac{12+18i}{13} \\
1 & \frac{-5+i}{13} & \frac{-5-i}{13} \\
1 & 1 & 1
\end{pmatrix}}^{-1}$

, και αφού πολλαπλασιάσουμε με την αρχική κατανομή $π_{1} = (1,0,0)$, παίρνουμε τελικά: $π_{20} = (0.040, 0.720, 0.240)$, όπου βλέποντας την πρώτη συντεταγμένη, είναι η πιθανότητα που θέλουμε, δηλαδή $(π_{20})_1 = \mathbb{P}[X_{20} = 1 ]= 0.040$ , αποτέλεσμα που συμβαδίζει με τα υπολογιστικά μας αποτελέσματά.

---
3. Με χρήση τώρα της εντολής `stat.variance(p_list)`, μπορούμε να υπολογίσουμε τη δειγματική διασπορά των εκτιμήσεων για $N=200$ και $N=20.000$, και τις βρίσκουμε ίσες με : $S^2(\hat{p}_{200})=3 \times 10^{-5}$ και $S^2(\hat{p}_{20000})=0.1 \times 10^{-5}$
---
4. Αν η αλυσίδα ξεκινήσει από την κατάσταση 3, βλέπουμε από τον πίνακα πιθανοτήτων μετάβασης πως η πιθανότητα να μεταβεί στην κατάσταση 2 (όπως κάνει και αν ξεκινήσει από την 1) είναι $5/6$ , άρα σύμφωνα με τη μαρκοβιανή ιδιότητα, η αλυσίδα μας από την κατάσταση 2 και έπειτα συμπεριφέρεται σαν να είχε ξεκινήσει από την 1. Αυτό αποδεικνύεται και πειραματικά, αφού τα αποτελέσματα, για αρχή από την 3 είναι τα ίδια όπως αυτά για αρχή από την 1, δηλαδή :$ \mathbb{P}\left[X_{20} = 1 \mid X_0 = 1 \right]=\mathbb{P}\left[X_{40} = 1 \mid X_0 = 3 \right]$ αφού $\mathbb{P}\left[X_{1} = 2 \mid X_0 = 3 \right]=5/6 $ .

---

# Άσκηση 40

   Στην συγκεκριμένη άσκηση θα υπολογίσουμε με τον ίδιο τρόπο (δειγματικό μέσο όρο των εκτιμητριών _Monte Carlo_ ), με $M=50$ δείγματα, την πιθανότητα νίκης του παίκτη που σερβίρει πρώτος( `PlayerA` ) σε ένα παιχνίδι Tennis, δεδομένο ότι η πιθανότητα πόντου είναι $p=0.6$ .
   Ο πίνακας πιθανοτήτων μετάβασης αντλήθηκε από την εικόνα της άσκησης.

In [2]:
from simple_markov_chain_lib import markov_chain
import statistics as stat

p = .6  #Probability of PlayerA scoring a point

markov_table = {    # Markov Table of the 17 states of a Tennis game
                    # based on the image
    '0-0': {'15-0': p, '0-15': 1 - p},  # 1st row of the pic

    '15-0': {'30-0': p, '15-15': 1 - p}, '0-15': {'15-15': p, '0-30': 1 - p},  # 2nd row

    '30-0': {'40-0': p, '30-15': 1 - p}, '15-15': {'30-15': p, '15-30': 1 - p},  # 3rd row
    '0-30': {'15-30': p, '0-40': 1 - p},

    '40-0': {'GameA': p, '40-15': 1 - p}, '30-15': {'40-15': p, 'Deuce': 1 - p},  # 4th row
    '15-30': {'Deuce': p, '15-40': 1 - p}, '0-40': {'15-40': p, 'GameB': 1 - p},

    '40-15': {'GameA': p, 'AdvA': 1 - p}, 'Deuce': {'AdvA': p, 'AdvB': 1 - p}, '15-40': {'AdvB': p, 'GameB': 1 - p},
    # 5th row

    'GameA': {'GameA': 1.}, 'AdvA': {'GameA': p, 'Deuce': 1 - p},  # 6th row
    'AdvB': {'Deuce': p, 'GameB': 1 - p}, 'GameB': {'GameB': 1.}
}

initial_dist = {'0-0': 1.}  # Every game starts from 0-0

# Markov Chain construction
mc = markov_chain(markov_table, initial_dist)

# Experiment parameters
N: int = 10000   # number of samples
M: int = 50     # Sample size
p_list = []     # List of the phat estimates


# Simulation
for j in range(M):
    counter = 0
    for i in range(N):
        mc.start()  # new experiment
        while True:
            mc.move()
            if mc.running_state == 'GameA':
                counter += 1
                break
            elif mc.running_state == 'GameB':
                break
    
    phat: float = counter / N
    p_list.append(phat)

print('The probability of PlayerA winning is {:.2%}. '.format(stat.mean(p_list)))
# The code to compute the sample mean of our phat list AKA the probability of PlayerA winning

The probability of PlayerA winning is 73.58%. 


Άρα, υπολογίσαμε την δειγματική μέση τιμή, δείγματους μήκους 50, της εκτιμήτριας _Monte Carlo_ να είναι : 

$\overline{\hat{p}_{10000}}=74\%$ , ή αλλίως την πιθανότητα νίκης του `PlayerA` .